pulser/simulation/simulation.py

```diff
@@ -297,57 +337,222 @@ def get_hamiltonian(self, time):
+            if hasattr(self._seq, '_measurement'):
+                return self._seq._measurement
             else:
-                if initial_state.shape != (self.dim**self._size,):
-                    raise ValueError("Incompatible shape of initial_state")
-                self._initial_state = qutip.Qobj(initial_state)
-        else:
-            # by default, initial state is "ground" state of g-r basis.
-            all_ground = [self.basis['g'] for _ in range(self._size)]
-            self._initial_state = qutip.tensor(all_ground)
-
-        result = qutip.sesolve(self._hamiltonian,
-                               self._initial_state,
-                               self.eval_times,
-                               progress_bar=progress_bar,
-                               options=qutip.Options(max_step=5,
-                                                     **options)
-                               )
-
-        if hasattr(self._seq, '_measurement'):
-            meas_basis = self._seq._measurement
+                if self.basis_name in {'digital', 'all'}:
+                    return 'digital'
+                else:
+                    return 'ground-rydberg'
+
+        def _run_solver(as_subroutine=False,
+                        measurement_basis=None):
+            """Returns:
+                CleanResults: Object containing the time evolution results. Its
+                _states attribute contains QuTiP quantum states, not Counters.
+            """
+            if not as_subroutine:
+                # CLEAN SIMULATION:
+                # Build (clean) hamiltonian
+                self._build_hamiltonian()
+                measurement_basis = _assign_meas_basis()
+
+            result = qutip.mesolve(self._hamiltonian,
+                                   self._config['initial_state'],
+                                   self.eval_times,
+                                   c_ops=self._collapse_ops,
+                                   progress_bar=progress_bar,
+                                   options=qutip.Options(max_step=5,
+                                                         **options)
+                                   )
+            return CleanResults(result.states, self.dim, self._size,
+                                self.basis_name, measurement_basis)
+
+        if self._noise:
+            # NOISY SIMULATION:
+            # We run the system multiple times
+            time_indices = range(len(self.eval_times))
+            total_count = np.array([Counter() for _ in time_indices])
+            for _ in range(self._config['runs']):
+                # At each run, new random noise
+                if 'SPAM' in self._noise:
+                    self._prepare_spam_detune()
+                self._build_hamiltonian()
+                meas_basis = _assign_meas_basis()
+                # Get CleanResults instance from sequence with added noise:
+                res_with_noise = _run_solver(as_subroutine=True,
```

Comment:
It is a bit awkward to call this `res_with_noise` although it is a `CleanResults` instance. Perhaps something like `clean_res_noisy_seq` would be more explicit to read (and would make clear the comment above)